File: backend/app/services/scoring.py
import math

from app.models import Hospital, HospitalWithScore, SortMode
from app.services.maps import TravelResult


def compute_scores(
    hospitals: list[Hospital],
    travel_results: list[TravelResult],
    sort_by: SortMode = SortMode.shortest_total,
    wait_weight: float = 0.5,
) -> list[HospitalWithScore]:
    # Build scored list with all computed fields
    scored: list[HospitalWithScore] = []
    for hospital, travel in zip(hospitals, travel_results):
        wait = hospital.wait_time_minutes or 240  # default 4h if unknown
        total = wait + travel.duration_minutes

        scored.append(
            HospitalWithScore(
                **hospital.model_dump(),
                travel_time_minutes=travel.duration_minutes,
                distance_km=travel.distance_km,
                total_time_minutes=total,
                priority_score=0,
                rank=0,
            )
        )

    # Normalize for custom weighting
    max_wait = max((h.wait_time_minutes or 240) for h in hospitals) or 1
    max_travel = max(t.duration_minutes for t in travel_results) or 1

    for h in scored:
        wait = h.wait_time_minutes or 240
        norm_wait = wait / max_wait
        norm_travel = h.travel_time_minutes / max_travel

        if sort_by == SortMode.shortest_wait:
            h.priority_score = round(wait, 1)
        elif sort_by == SortMode.shortest_commute:
            h.priority_score = round(h.travel_time_minutes, 1)
        elif sort_by == SortMode.shortest_total:
            h.priority_score = round(h.total_time_minutes, 1)
        else:
            # Custom: use wait_weight slider (0 = all commute, 1 = all wait)
            travel_weight = 1.0 - wait_weight
            score = (
                wait_weight * math.log1p(norm_wait * 10)
                + travel_weight * math.log1p(norm_travel * 10)
            )
            h.priority_score = round(score, 2)

    scored.sort(key=lambda h: h.priority_score)
    for i, h in enumerate(scored):
        h.rank = i + 1

    return scored

File: src/lib/scoring.ts
import type { Hospital, HospitalWithScore } from "@/types/hospital";
import type { TravelResult } from "@/lib/maps";

interface ScoringWeights {
  waitTime: number;
  travelTime: number;
}

const SEVERITY_WEIGHTS: Record<string, ScoringWeights> = {
  low: { waitTime: 0.3, travelTime: 0.7 },      // Prefer closer hospitals
  medium: { waitTime: 0.5, travelTime: 0.5 },    // Balanced
  high: { waitTime: 0.7, travelTime: 0.3 },      // Prefer shorter waits
  critical: { waitTime: 0.9, travelTime: 0.1 },  // Nearest capable hospital
};

export function computeScores(
  hospitals: Hospital[],
  travelResults: TravelResult[],
  severity: string = "medium"
): HospitalWithScore[] {
  const weights = SEVERITY_WEIGHTS[severity] || SEVERITY_WEIGHTS.medium;

  // Normalize values for scoring
  const maxWait = Math.max(
    ...hospitals.map((h) => h.waitTimeMinutes ?? 240)
  );
  const maxTravel = Math.max(...travelResults.map((t) => t.durationMinutes));

  const scored: HospitalWithScore[] = hospitals.map((hospital, i) => {
    const travel = travelResults[i];
    const waitTime = hospital.waitTimeMinutes ?? 240; // Default to 4h if unknown

    // Normalized scores (0 = best, 1 = worst)
    const normalizedWait = maxWait > 0 ? waitTime / maxWait : 0;
    const normalizedTravel =
      maxTravel > 0 ? travel.durationMinutes / maxTravel : 0;

    // Combined score using logarithmic scaling for diminishing returns
    const waitScore = Math.log1p(normalizedWait * 10);
    const travelScore = Math.log1p(normalizedTravel * 10);

    const priorityScore =
      weights.waitTime * waitScore + weights.travelTime * travelScore;

    return {
      ...hospital,
      travelTimeMinutes: travel.durationMinutes,
      distanceKm: travel.distanceKm,
      priorityScore: Math.round(priorityScore * 100) / 100,
      rank: 0,
    };
  });

  // Sort by priority score (lower is better)
  scored.sort((a, b) => a.priorityScore - b.priorityScore);

  // Assign ranks
  scored.forEach((h, i) => {
    h.rank = i + 1;
  });

  return scored;
}
